astropy/stats/bayesian_blocks.py

```diff
@@ -286,24 +277,22 @@ def p0_prior(self, N, Ntot):
         """
         return 4 - np.log(73.53 * self.p0 * (N ** -0.478))
 
-    def gamma_prior(self, N, Ntot):
-        """
-        Basic prior, parametrized by slope gamma.
-        See eq. 3 in Scargle (2012)
-        """
-        if self.gamma == 1:
-            return 0
-        else:
-            return (np.log(1 - self.gamma)
-                    - np.log(1 - self.gamma ** (Ntot + 1))
-                    + N * np.log(self.gamma))
-
     # the fitness_args property will return the list of arguments accepted by
     # the method fitness().  This allows more efficient computation below.
     @property
     def _fitness_args(self):
         return signature(self.fitness).parameters.keys()
 
+    def compute_ncp_prior(self, N):
+        """
+        If ``ncp_prior`` is not explicitly defined, compute it from ``gamma``
+        or ``p0``.
+        """
+        if self.gamma is not None:
+            self.ncp_prior = -np.log(self.gamma)
```

Comment:
Hi … this is a neat idea.  I thought about it in the past,
but I guess the constant (ncp_prior) idea worked so well, at
least for the problems I dealt with, that I never went into it.
And I never ran into motivation for specific alternative forms
for the prior on the number of blocks.  But that is no reason
not to allow this freedom if it is practical.

However, the need to preserve block additivity of the
fitness function -- eq. (7) of our 2012 paper , and necessary
for the logic of the algorithm -- limits what form the prior can
have.  So it is not clear to me that you can allow user-defined
priors at all.

If there is some way to work around this issue, don't
forget that in addition to saving the optimum fitness
values for the previous solutions you would have to save the
number of blocks in the solutions.  I assume this would be
straightforward.

All of this would not really impact the computation time
very much, would it?

Bravo for pursuing these ideas!  In addition one might
think about more explicit constraints.  E.g., set a maximum
on the number of blocks … or force a range between some
minimum and maximum, with an arbitrary distribution
in between … or maybe only allow an even number of blocks …

Going even farther out into left field, forcing the amplitude of
the first  and last blocks to be equal is something that might
be of use in connection with the "data on the circle" problem,
or say in GRB analysis where you know the background rate
before and after the burst is the same.

But I never found a way to implement these ideas, other than
the way we solved the circular data problem.

Cheers, Jeff

Jeffrey D. Scargle
Space Science Division
Mail Stop 245-3
NASA Ames Research Center
Moffett Field, CA 94035-0001

Phone: 650 604 6330
Cell:  415 385 4297
Fax: 650 604 6779
________________________________
From: robcross [notifications@github.com]
Sent: Thursday, November 19, 2015 9:39 PM
To: astropy/astropy
Cc: Scargle, Jeffrey D. (ARC-SST)
Subject: Re: [astropy] Fixes implementation of prior function in Bayesian Blocks algorithm (#4301)


In astropy/stats/bayesian_blocks.py<https://github.com/astropy/astropy/pull/4301#discussion_r45437490>:

>      # the fitness_args property will return the list of arguments accepted by
>      # the method fitness().  This allows more efficient computation below.
>      @property
>      def _fitness_args(self):
>          return signature(self.fitness).parameters.keys()
>
> +    def compute_ncp_prior(self, N):
> +        """
> +        If ``ncp_prior`` is not explicitly defined, compute it from ``gamma``
> +        or ``p0``.
> +        """
> +        if self.gamma is not None:
> +            self.ncp_prior = -np.log(self.gamma)


Since time is tight right now, what do you think about leaving this as is for now?

I'm thinking of how to let the user specify a prior, and I want to make sure I do it right.

The reason that Scargle chose the geometrical prior is that the math works out so well.
If a user wants an arbitrary function of N_blocks, we could allow them to subclass something like PriorFunc, and define a general p(N_blocks). Then in place of ncp_prior, we could take ln(p(N_blocks+1)) - ln(p(N_blocks)). This would be slow, but if there's a simple closed form of ln(p(N_blocks)), we could allow the user to overload that instead, which would be faster. Just lnp(N_blocks+1) - lnp(N_blocks).

I'm not sure how sound this is. I've fixed all the issues brought up except for this one for now. I'll push what I have for now.

—
Reply to this email directly or view it on GitHub<https://github.com/astropy/astropy/pull/4301/files#r45437490>.